### DISCUSSION TITLE
Using CoreWCF as client for external SOAP API with multiple Domains

### DISCUSSION THREAD

**Discussion by JonathanCruse** [CoreWCF/CoreWCF#1363] [Q&A]
Hello there,

I have an external system which API I need to access through basic authentication . 
My basic client is working fine - either by using ChannelFactory or by directly using PortClient.

Problem: The external system has multiple Domains (different customers can have unique domains)

Is there a way to change the URL, where my POST-requests are sent without connecting the different APIs as separate referenced services?

What did I try?
1. `client.Endpoint.Address = new System.ServiceModel.EndpointAddress($"https://{Domain}/soap.php/soap/container")` This told me, that the requested service, e.g. "getAll" is not present. I assume, that the EndpointAddress only searches for the matching data annotation attributes in the reference.cs file
2. I could include all references for all domains (they are growing tho) and assign everything dynamically. I was able to make it work in a prototype, but it is really messy and insanely hard to debug if something goes wrong.

Is there a suggested way to to that ?

An alternative solution could be to use HttpClient instead and do requests "manually" - is there a way to only use de-/serialisation of WcfCore?

Thanks!
Jonathan

**Comment by mconnew** [comment]
Client issues should be submitted to https://github.com/dotnet/wcf. Any more follow ups should be done in a new issue there.  
The solution you are looking for is to create an instance of ChannelFactory using the constructor overload which only takes a Binding and not an EndpointAddress. Then create a channel using the CreateChannel overload which takes an EndpointAddress. This way a single ChannelFactory instance can create channels for multiple service endpoints. If you have different credentials for each endpoint, then there are multiple solutions with different levels of complexity and abstraction to solve your problem, depending on your exact scenario. Are you always using basic auth? Do you know the set of endpoints and corresponding credentials at the time of creating the ChannelFactory? If not, do you know the credentials before you need to create the channel? What's the process for obtaining the credentials for an endpoint? Discussion about the credentials needs to continue on the dotnet/wcf repo, but these are the things needed to be known to suggest a solution.

**Reply by JonathanCruse** [reply]
Thanks for your extensive answer. I'll look into those things and if I cannot solve it, I'll direct the question to the wcf repository. Best, Jonathan
